File: 310971/Stankevich/rest310/src/main/java/com/task/rest/service/NoticeService.java
package com.task.rest.service;

import com.task.rest.dto.NoticeRequestTo;
import com.task.rest.dto.NoticeResponseTo;
import com.task.rest.exception.ResourceNotFoundException;
import com.task.rest.mapper.NoticeMapper;
import com.task.rest.model.Notice;
import com.task.rest.repository.NoticeRepository;
import com.task.rest.repository.TweetRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class NoticeService {

    private final NoticeRepository repository;
    private final TweetRepository tweetRepository;
    private final NoticeMapper mapper;

    public NoticeResponseTo create(NoticeRequestTo request) {
        if (!tweetRepository.existsById(request.getTweetId())) {
            throw new ResourceNotFoundException(
                    "Tweet not found with id: " + request.getTweetId(), "40411");
        }

        Notice notice = mapper.toEntity(request);
        notice = repository.save(notice);
        return mapper.toDto(notice);
    }

    public NoticeResponseTo getById(Long id) {
        Notice notice = repository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException(
                        "Notice not found with id: " + id, "40431"));
        return mapper.toDto(notice);
    }

    public List<NoticeResponseTo> getAll() {
        return mapper.toDtoList(repository.findAll());
    }

    public NoticeResponseTo update(NoticeRequestTo request) {
        if (request.getId() == null || !repository.existsById(request.getId())) {
            throw new ResourceNotFoundException(
                    "Notice not found with id: " + request.getId(), "40432");
        }

        if (!tweetRepository.existsById(request.getTweetId())) {
            throw new ResourceNotFoundException(
                    "Tweet not found with id: " + request.getTweetId(), "40411");
        }

        Notice notice = mapper.toEntity(request);
        notice = repository.update(notice);
        return mapper.toDto(notice);
    }

    public void delete(Long id) {
        if (!repository.existsById(id)) {
            throw new ResourceNotFoundException(
                    "Notice not found with id: " + id, "40433");
        }
        repository.deleteById(id);
    }

    public List<NoticeResponseTo> getByTweetId(Long tweetId) {
        if (!tweetRepository.existsById(tweetId)) {
            throw new ResourceNotFoundException(
                    "Tweet not found with id: " + tweetId, "40411");
        }
        return mapper.toDtoList(repository.findByTweetId(tweetId));
    }
}

File: 351003/brezgunov/distcomp/src/main/kotlin/com/distcomp/service/NoticeService.kt
package com.distcomp.service

import com.distcomp.dto.kafka.KafkaNoticeRequest
import com.distcomp.dto.kafka.KafkaNoticeResponse
import com.distcomp.dto.notice.NoticeRequestTo
import com.distcomp.dto.notice.NoticeResponseTo
import com.distcomp.exception.NewsNotFoundException
import com.distcomp.repository.NewsRepository
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.KotlinModule
import org.springframework.stereotype.Service
import java.util.UUID
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

@Service
class NoticeService(
    private val producer: NoticeProducer,
    private val registry: KafkaResponseRegistry,
    private val newsRepository: NewsRepository
) {
    private val mapper = ObjectMapper().registerModule(KotlinModule.Builder().build())

    fun getById(id: Long): NoticeResponseTo {
        val response = sendAndWait(KafkaNoticeRequest(id = id.toString(), type = "GET"))
        return convertToSingleDto(response.data)
    }

    fun getAll(): List<NoticeResponseTo> {
        val response = sendAndWait(KafkaNoticeRequest(id = "all", type = "GET_ALL"))
        return convertToList(response.data)
    }

    fun createNotice(noticeRequestTo: NoticeRequestTo): NoticeResponseTo {
        if (!newsRepository.existsById(noticeRequestTo.newsId)) {
            throw NewsNotFoundException("News not found")
        }
        val response = sendAndWait(KafkaNoticeRequest(id = UUID.randomUUID().toString(), type = "POST", payload = noticeRequestTo))
        return convertToSingleDto(response.data)
    }

    fun updateNotice(noticeId: Long, noticeRequestTo: NoticeRequestTo): NoticeResponseTo {
        if (!newsRepository.existsById(noticeRequestTo.newsId)) {
            throw NewsNotFoundException("News not found")
        }
        val response = sendAndWait(KafkaNoticeRequest(id = noticeId.toString(), type = "PUT", payload = noticeRequestTo))
        return convertToSingleDto(response.data)
    }

    fun deleteById(id: Long) {
        sendAndWait(KafkaNoticeRequest(id = id.toString(), type = "DELETE"))
    }

    private fun sendAndWait(request: KafkaNoticeRequest): KafkaNoticeResponse {
        val future = registry.register(request.id)
        producer.send(request)
        return try {
            future.get(1, TimeUnit.SECONDS)
        } catch (_: TimeoutException) {
            throw RuntimeException("timeout")
        }
    }

    private fun convertToSingleDto(value: Any?): NoticeResponseTo =
        mapper.convertValue(value, NoticeResponseTo::class.java)

    private fun convertToList(value: Any?): List<NoticeResponseTo> =
        mapper.convertValue(value, object : TypeReference<List<NoticeResponseTo>>() {})
}